fx_compile.py

```diff
@@ -0,0 +1,432 @@
+            load = te.ifThenElse(idxs[dim] < to_expr(sm), load, tensor.load(new_idxs))
+        return load
+    return te.Compute(name, get_dim_args(out_shape), f)
+
+lowering_functions[torch.transpose] = transpose_lower
+lowering_functions[torch.flatten] = flatten_lower
+lowering_functions[torch.cat] = cat_lower
+
+def bmm_lower(name, out_shape, inp_shapes, args):
+    M1 = args[0]
+    M2 = args[1]
+    B, N, M = inp_shapes[0][0]
+    P = inp_shapes[1][0][2]
+
+    def f(b, n, p, m):
+        return M1.load([b, n, m]) * M2.load([b, m, p])
+    mm = te.Compute('mm', get_dim_args([B,N,P,M]), f)
+    return te.SumReduce(name, get_dim_args([B, N, P]), mm, get_dim_args([M]))
+
+
+def mm_lower(name, out_shape, inp_shapes, args):
+    M1 = args[0]
+    M2 = args[1]
+    N, M = inp_shapes[0][0]
+    P = inp_shapes[1][0][1]
+
+    def f(n, p, m):
+        return M1.load([n, m]) * M2.load([m, p])
+    mm = te.Compute('mm', get_dim_args([N,P,M]), f)
+    return te.SumReduce(name, get_dim_args([N, P]), mm, get_dim_args([M]))
+
+lowering_functions[torch.bmm] = bmm_lower
+lowering_functions[torch.mm] = mm_lower
+
+
+
+def lower_function(name, op, shape_env, nnc_args, args):
+    inp_shapes = fx.node.map_aggregate(args, lambda arg: shape_env[arg.name] if isinstance(arg, fx.Node) else None)
+    return lowering_functions[op](name, shape_env[name][0], inp_shapes, nnc_args)
+
+def nnc_compile(model: torch.nn.Module, shape_specialize, shape_env) -> torch.nn.Module:
+    fx_model = fx.symbolic_trace(model)
+    env = {}
+    def get_te_shapes(name):
+        return [te.ExprHandle.int(i) for i in shape_env[name][0]]
+
+    def get_nnc_type(dtype):
+        if dtype == torch.float:
+            return torch._C.te.Dtype.Float
+        elif dtype == torch.long:
+            return torch._C.te.Dtype.Long
+        else:
+            raise RuntimeError("nyi")
+
+    def get_te_type(name):
+        return get_nnc_type(shape_env[name][1])
+
+    def gen_compute(args):
+        te_args = [env[arg.name] for arg in args]
+
+    def lookup_env(l):
+        return fx.node.map_aggregate(l, lambda x: env[x.name] if isinstance(x, fx.Node) else x)
+
+    def fetch_attr(target : str):
+        target_atoms = target.split('.')
+        attr_itr = fx_model
+        for i, atom in enumerate(target_atoms):
+            if not hasattr(attr_itr, atom):
+                raise RuntimeError(f"Node referenced nonexistant target {'.'.join(target_atoms[:i])}")
+            attr_itr = getattr(attr_itr, atom)
+        return attr_itr
+
+    outs = None
+    inputs = []
+    module_attrs = []
+    for node in fx_model.graph.nodes:
+        if node.op == 'placeholder':
+            shapes = get_te_shapes(node.name)
+            env[node.name] = te.Placeholder(node.name, get_te_type(node.name), shapes)
+            inputs.append(env[node.name])
+        elif node.op == 'call_function':
+            result = lower_function(node.name, node.target, shape_env, lookup_env(node.args), node.args)
```

Comment:
Same sort of comment as above, this will miss callsites with some arguments as kwargs. If you're just handling functions, you might be able to use something like `inspect.Signature.bind()` to normalize args/kwargs